primenet.py

```diff
@@ -4446,33 +4786,49 @@ def get_assignments(adapter, adir, cpu_num, progress, tasks):
     _percent = time_left = None
     if progress is not None:
         _percent, time_left, _, _ = progress  # unpack update_progress output
-    num_cache = options.num_cache + 1
+    num_cache = options.num_cache
+    if not num_cache:
+        num_cache = 10 if options.mfaktc or options.mfakto else 1
+        adapter.debug("Config option num_cache wasn't set, defaulting to {0:n} assignment{1}".format(num_cache, "s" if num_cache != 1 else ""))
     if options.password:
         num_cache += 1
     num_existing = len(assignments)
+    if num_existing > num_cache:
+        adapter.debug(
+            "Number of existing assignments ({0:n}) was higher than num_cache ({1:n}), num_cache increased to {0:n}".format(
+            num_existing, num_cache))
+        num_cache = num_existing
     num_cache = max(num_existing, num_cache)
     if time_left is not None:
         time_left = timedelta(seconds=time_left)
         days_work = timedelta(days=options.days_of_work)
         if time_left <= days_work:
-            num_cache += 1
+            num_cache = math.ceil(num_existing * days_work / time_left)
             adapter.debug(
-                "Time left is {0} and smaller than days_of_work ({1}), so num_cache is increased by one to {2:n}".format(
+                "Time left is {0} and smaller than days_of_work ({1}), so num_cache increased to {2:n}".format(
                     time_left, days_work, num_cache
                 )
             )
-    amax = config.getint(SEC.PrimeNet, "MaxExponents") if config.has_option(SEC.PrimeNet, "MaxExponents") else 15
-    num_cache = min(num_cache, amax)
+    else:
+        adapter.debug(
+            "No time estimate available for current {0:n} assignment{1}, so we'll only fetch {2:n} for now, instead of {3:n} day{4} of work".format(
+                num_existing, "s" if num_existing != 1 else "", num_cache, options.days_of_work, "s" if options.days_of_work != 1 else ""))
+    if config.has_option(SEC.PrimeNet, "MaxExponents"):
+        amax = config.getint(SEC.PrimeNet, "MaxExponents")
+        if amax < num_cache:
+            adapter.debug(
+                "Config option MaxExponents ({0:n}) is smaller than num_cache ({1:n}), num_cache decreased to {0:n}".format(
+                    amax, num_cache, workfile))
+            num_cache = amax
```

Comment:
If only to prevent a bug in the program from causing it to request too many assignments, I would prefer to keep a default value for the `MaxExponents` config file option, even if it is significantly larger when the user is doing TF assignments.

Reply:
On one level I agree, as it will surely happen that someone gets way too many assignments due to a bug or misunderstanding if we don't include some safeguard, but I just personally found the experience of having to discover the default MaxExponents option in the code to be a bad user experience.

There are several factors that contribute to the confusion in the previous version:
1. there is no cli option for max exponents, so -h doesn't lead you in the right direction
2. it's not mentioned in --setup
3. there's a cli option for "max exponent" but not max exponents (two different options that are confusingly similar)
4. in the previous configuration, even if you have no MaxExponents in the config (which happens by default), you're still limited to 15, so you have to search through the actual code to figure out what's going on. (even if we raise this, it will still happen when a user puts in a high days_of_work for certain worktypes)
5. throughout this whole process, you're not actually told about MaxExponents in the log

I have added some debug messages to hopefully make it clearer if this is happening to someone in the future (5), but they would need to have debugging on, which, I think if they're having to use debugging to figure out why the program isn't behaving like they told it to (using days work), we've already failed some test of the user experience.

10 days of TF work for my machine can be ~1500 assignments at small (but still ftc wavefront) bitlevels, which is an enormous amount of exponents, and the maximum number of days of work to fetch is 90, which would be ~9470 more averagely sized assignments for me.

I think I heard 10,000 was the maximum number of assignments you could have in primenet, though if I grabbed 10,000 wavefront TF assignments that would likely greatly negatively impact the project. We surely want the limit to be below this, but still respect days_of_work


TL;DR: I think we should have safeguards but they should be upfront.

Proposal:
We could add it to --setup with an appropriately sized default for the worktype. (People become aware of its existence and set it according to their comfort level)
When creating the .ini file, we put that default in (even if the user doesn't use --setup), so when they're editing the .ini they will see it and realize that's what's keeping the script from getting more assignments
We can use the previously mentioned default if they delete the line in .ini, but display a warning that doesn't require the debug verbosity to see, that we are using max_exponents=15 or 100 (or something) as a default. (Alternatively, just make all the num_cache related debugs into `info` messages)

Reply:
Thanks for the detailed write up.

The `MaxExponents` config file option comes from Prime95/MPrime, which also defaults to 15 assignments. Note that this limit is per worker, so with multiple workers one can have many more assignments. I was hesitant at first to add support for this option, but over time I became concerned that a bug either in the script or on the server could cause the program to request too many assignments, so I decided to add this as a safeguard. Prime95/MPrime of course goes a step further and will actually automatically unreserve assignments, which I would never do.

Anyway, at least with CPUs, 15 wavefront assignments is typically a lot of work (days or weeks), especially with primality tests, so most users should never hit this limit. However, GPUs can obviously be faster, so feel free to increase the limit for TF work to whatever value makes sense, even if that is hundreds or thousands of assignments. Ideally, only the most advanced users should ever need to change this option. I believe TF (and P-1) assignments are only valid for 30 days, so requesting 90 days of work would not be applicable here.

Another problem I am aware of is that there is no exhaustive list of all supported config file options (without reading my release notes for each version). It would be easy enough to create such a list, but I am not sure the best place to store it. Maybe on the README? Long term, we probably need to write more extensive documentation on how to use the program.

Regarding your proposal, adding a log message when the `MaxExponents` value is hit was a good improvement, so thanks for implementing that. Note that in general I do not want to script to be too verbose by default, as users may miss important messages. I wish the Python logging module supported multiple levels of debug messages instead of the `critical` level, which is currently unused, but alas all we have is `debug`. Feel free to put a default value for `MaxExponents` in the `local.ini` config file. In that case, you might also restore the `results_file` as well. I would prefer to not add a prompt to the `--setup` option, as users would then just set a large value defeating the purpose of this feature, but I am open to being convinced otherwise.

Reply:
> The MaxExponents config file option comes from Prime95/MPrime, which also defaults to 15 assignments. 

Ah, didn't realize this was directly from p95, that makes a bit more sense then.

>  I believe TF (and P-1) assignments are only valid for 30 days, so requesting 90 days of work would not be applicable here.

This is a good point, I think non Cat 0 - 1 exponents are set to 60 days by default, mine mostly shows 60 for all of my assignment types here: https://www.mersenne.org/manual_extension/
Though I manually extended some P-1 jobs that I'm doing jointly and they now say 218 days.

Either way, 1000 TF assignments is certainly enough for most people, if you have more than 100 assignments you get put on a watchlist anyway. (https://www.mersenne.org/assignments/topusers.php)

Though I think this list actually uses 1000 for the TF threshold, otherwise I would be on it.

> Feel free to put a default value for MaxExponents in the local.ini config file.

Can do!

> In that case, you might also restore the results_file as well.

I'm not sure what you mean by restoring the results_file, though it may come to me while I'm in the code.

> I would prefer to not add a prompt to the --setup option, as users would then just set a large value defeating the purpose of this feature, but I am open to being convinced otherwise.

That's fair. I suppose gating getting an obscene amount of assignments behind config editing is pretty good. I was envisioning it as a "launch control lock switch" sort of deal, where the user is allowed to do whatever they want with the program, but we make it difficult to do on accident.

Reply:
Just spitballing, but maybe we have the MaxExponents default to some value based on their selected days of work and their worktype? This way we guard against wildly inaccurate/buggy time estimates causing lots of work being fetched, but we stay out of the way of users that actually want that much work? For TF this can maybe be like 150 assignments per day, I'm not sure what other worktypes should default to though.

Reply:
> Ah, didn't realize this was directly from p95, that makes a bit more sense then.

Yes, for just a general hint, the camel cased config file options are either the same or adapted from a similar option in Prime95/MPrime, while the snake cased ones are unique to the PrimeNet program.

> I'm not sure what you mean by restoring the results_file, though it may come to me while I'm in the code.

When we removed the default value for the `--resultsfile` option, this also removed it from the config file by default. I was just proposing that you manually add it back in the config file when you add the default value for the `MaxExponents` option.

> I was envisioning it as a "launch control lock switch" sort of deal, where the user is allowed to do whatever they want with the program, but we make it difficult to do on accident.

Yes, I agree, that was my original goal with this feature. I am all for empowering power users to do what they need to do.

> Just spitballing, but maybe we have the MaxExponents default to some value based on their selected days of work and their worktype?

This would be fine with me, but it would of course be more difficult to implement, as we could then no longer have a hard coded default in the config file. I was previously going for simplicity.

Reply:
All these should be done now!